### PR TITLE
Add coordinator book appointment function

### PR DESCRIPTION
**Pull request by yedidya03**

### CODE REVIEW

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  MANUAL_DONOR_ID,
+} from "@zm-blood-components/common";
+import { dbAppointmentToBookedAppointmentApiEntry } from "../utils/ApiEntriesConversionUtils";
+import { notifyOnAppointmentBooked } from "../notifications/BookAppointmentNotifier";
+import { BookAppointmentStatus } from "../../../common/src/functions-api";
+import { DbDonor } from "../function-types";
+
+const WEEKS_BUFFER = 0;
+
+export default async function (
```
Comment (yaron1m) [inline packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:22]: How much of this code is duplicate from the BookAppointmentHandler?

Can you extract common code to utils and prevent duplications?

File: packages/common/src/functions-api.ts [hunk 2]
```diff
@@ -23,6 +24,8 @@ export type AvailableAppointmentApiEntry = {
 
 export type BookedAppointmentApiEntry = {
   id: string;
+  donorDetails?: MinimalDonorDetailsForAppointment;
+  assigningCoordinatorId?: string;
```
[the remark this answers is not in the document]
Comment (yedidya03) [reply packages/common/src/functions-api.ts:28]: Done

File: packages/common/src/functions-api.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { MinimalDonorDetailsForAppointment } from ".";
```
[the remark this answers is not in the document]
Comment (yedidya03) [reply packages/common/src/functions-api.ts:1]: Done

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.test.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+import * as admin from "firebase-admin";
+import { sampleUser } from "../testUtils/TestSamples";
+import { notifyOnAppointmentBooked } from "../notifications/BookAppointmentNotifier";
+import { mocked } from "ts-jest/utils";
+import { BookAppointmentStatus } from "../../../common/src/functions-api";
+import {
+  AppointmentStatus,
+  MANUAL_DONOR_ID,
+  MinimalDonorDetailsForAppointment,
+} from "@zm-blood-components/common/src";
```
[the remark this answers is not in the document]
Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.test.ts:24]: Done

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.test.ts [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  deleteAppointmentsByIds,
+  getAppointmentsByIds,
+  setAppointment,
+} from "../dal/AppointmentDataAccessLayer";
+import { expectAsyncThrows } from "../testUtils/TestUtils";
+import * as admin from "firebase-admin";
+import { sampleUser } from "../testUtils/TestSamples";
+import { notifyOnAppointmentBooked } from "../notifications/BookAppointmentNotifier";
+import { mocked } from "ts-jest/utils";
+import { BookAppointmentStatus } from "../../../common/src/functions-api";
```
[the remark this answers is not in the document]
Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.test.ts:19]: Done

File: packages/functions/src/index.ts [hunk 6]
```diff
@@ -40,7 +40,7 @@ export const getBookedDonationsInHospital = handler(
 );
 
 // Donor
-export const bookAppointment = handler(bookAppointmentHandler);
+export const donorBookAppointment = handler(donorBookAppointmentHandler);
```
Comment (yaron1m) [inline packages/functions/src/index.ts:43]: You can't change the API name. 
This will break all old versions calling this API.

Comment (yedidya03) [reply packages/functions/src/index.ts]: fixed

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts (hunk 1 above)
Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:4]: moved to a common file. I wondered what to do with the tests..

File: packages/functions/src/index.ts (hunk 6 above)
Comment (yedidya03) [reply packages/functions/src/index.ts]: Fixed

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts (hunk 1 above)
Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:4]: done

File: packages/common/src/functions-api.ts [hunk 7]
```diff
@@ -62,17 +65,23 @@ export interface GetDonorAppointmentsResponse {
   futureAppointments: BookedAppointmentApiEntry[];
 }
 
-export const BookAppointmentFunctionName = "bookAppointment";
+export const CoordinatorBookAppointmentFunctionName =
+  "coordinatorBookAppointment";
+export const DonorBookAppointmentFunctionName = "bookAppointment";
```
Comment (yaron1m) [inline packages/common/src/functions-api.ts:70]: Move this down next to the function request object

File: packages/common/src/functions-api.ts [hunk 8]
```diff
@@ -62,17 +65,23 @@ export interface GetDonorAppointmentsResponse {
...
+export interface BookAppointmentRequest {
```
Comment (yaron1m) [inline packages/common/src/functions-api.ts:71]: Why are donors and coordinators use the same request object?
I think coordinators should have a separate request object for a separate function

File: packages/functions/src/common/BookAppointmentHelper.ts [hunk 9]
```diff
@@ -10,21 +10,31 @@ import {
...
 import { dbAppointmentToBookedAppointmentApiEntry } from "../utils/ApiEntriesConversionUtils";
 import { notifyOnAppointmentBooked } from "../notifications/BookAppointmentNotifier";
-import { BookAppointmentStatus } from "../../../common/src/functions-api";
+import { BookAppointmentStatus } from "@zm-blood-components/common/src/functions-api";
 import { DbDonor } from "../function-types";
 
 const WEEKS_BUFFER = 0;
 
-export default async function (
+export async function BookAppointment(
```
Comment (yaron1m) [inline packages/functions/src/common/BookAppointmentHelper.ts:22]: bookAppointment

File: packages/functions/src/common/BookAppointmentHelper.ts [hunk 10]
```diff
@@ -10,21 +10,31 @@ import {
   BookingChange,
   FunctionsApi,
   Hospital,
+  MANUAL_DONOR_ID,
 } from "@zm-blood-components/common";
 import { dbAppointmentToBookedAppointmentApiEntry } from "../utils/ApiEntriesConversionUtils";
 import { notifyOnAppointmentBooked } from "../notifications/BookAppointmentNotifier";
-import { BookAppointmentStatus } from "../../../common/src/functions-api";
+import { BookAppointmentStatus } from "@zm-blood-components/common/src/functions-api";
```
Comment (yaron1m) [inline packages/functions/src/common/BookAppointmentHelper.ts:17]: Use the `FunctionsApi` you already have imported in line 11

File: packages/functions/src/common/BookAppointmentHelper.ts [hunk 11]
```diff
@@ -10,21 +10,31 @@ import {
...
 const WEEKS_BUFFER = 0;
 
-export default async function (
+export async function BookAppointment(
   request: FunctionsApi.BookAppointmentRequest,
+  isDonor: boolean,
   callerId: string
 ): Promise<FunctionsApi.BookAppointmentResponse> {
-  const donorId = callerId;
+  let donorId = callerId;
```
Comment (yaron1m) [inline packages/functions/src/common/BookAppointmentHelper.ts:27]: This is confusing. 
Change the parameter name in line 25 to `donorId` and let the callers take care of this logic.
You can add another optional field called `bookingCoordinatorId` that will only apply to the new flow

Comment (yaron1m) [reply packages/functions/src/common/BookAppointmentHelper.ts:27]: This way you can also remove the `isDonor` and deduce the caller of this function by the existence of the `bookingCoordinatorId`

File: packages/functions/src/donor/DonorBookAppointmentHandler.test.ts [hunk 12]
```diff
@@ -57,6 +57,7 @@ test("Unauthenticated user throws exception", async () => {
 });
 
 test("Donor not found throws exception", async () => {
+  await saveAppointment(APPOINTMENT_TO_BOOK_2, false, 3);
```
Comment (yaron1m) [inline packages/functions/src/donor/DonorBookAppointmentHandler.test.ts:60]: Why is this needed?

File: packages/common/src/functions-api.ts (hunk 7 above)
Comment (yedidya03) [reply packages/common/src/functions-api.ts:70]: it is next to the request object

File: packages/common/src/functions-api.ts (hunk 8 above)
Comment (yedidya03) [reply packages/common/src/functions-api.ts:71]: done

File: packages/functions/src/common/BookAppointmentHelper.ts (hunk 9 above)
Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts:22]: done

File: packages/functions/src/common/BookAppointmentHelper.ts (hunk 10 above)
Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts:17]: done

File: packages/functions/src/common/BookAppointmentHelper.ts (hunk 11 above)
Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts:27]: done

File: packages/functions/src/donor/DonorBookAppointmentHandler.test.ts (hunk 12 above)
Comment (yedidya03) [reply packages/functions/src/donor/DonorBookAppointmentHandler.test.ts:60]: Because the function checks if the appointment is available before it checks the donor. And if will not set an appointment the function will return NO_SUCH_APPOINTMENTS.

File: packages/functions/src/common/BookAppointmentHelper.ts [hunk 13]
```diff
@@ -11,43 +11,35 @@ import {
...
-import { BookAppointmentStatus } from "@zm-blood-components/common/src/functions-api";
 import { DbDonor } from "../function-types";
 
 const WEEKS_BUFFER = 0;
 
-export async function BookAppointment(
-  request: FunctionsApi.BookAppointmentRequest,
+export async function bookAppointment(
+  donorId: string,
   isDonor: boolean,
```
Comment (yaron1m) [inline packages/functions/src/common/BookAppointmentHelper.ts:24]: Is this still needed?

**Review by yaron1m — APPROVED** [review]

File: packages/functions/src/common/BookAppointmentHelper.ts (hunk 13 above)
Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts:24]: what do you mean?

Comment (yaron1m) [reply packages/functions/src/common/BookAppointmentHelper.ts:24]: `isDonor == true` when `coordinatorId == undefined` - So you don't need both fields

Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts:24]: oh write

**Comment by yaron1m** [comment]
Any update here?

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts [hunk 14]
```diff
@@ -1,13 +1,31 @@
...
+  MANUAL_DONOR_ID,
+} from "@zm-blood-components/common";
 import { bookAppointment } from "../common/BookAppointmentHelper";
+import * as CoordinatorDAL from "../dal/AdminDataAccessLayer";
 
 export default async function (
   request: FunctionsApi.CoordinatorBookAppointmentRequest,
   callerId: string
 ): Promise<FunctionsApi.BookAppointmentResponse> {
+  if (request.donorId !== MANUAL_DONOR_ID) {
```
Comment (yaron1m) [inline packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:13]: I think we should throw here.

If (and when) we want to support this feature, let's do it in a different PR

Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: why in a different PR? all the functionality is already written here..

Comment (yaron1m) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: I don't think this is the right way to do it,
But more important, this PR is very big and has been open for over 14d.
We need to close it.

Manually marking a manual donor or a real donor are two different features, and they deserve different PRs

Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: but we already been throw all the developing process here. why to erase that work?

Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: and originally that was the hole purpose of this PR - to add the option of a manual donor

Comment (yaron1m) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: The requirement is to add a manual donor, a dummy one. Not an existing real user.

Assigning a real user requires more things, like notifying the users and more.

If you want to keep the work that was done here, move it to a branch so we could use it later

Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: its not a real one. this line is sais : "if this is a manual donor dont check if the coordinator has permissions to this donor, and return bookAppointment(...)"

Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: the manual donors details stays in the appointment details, as planed..

Comment (yaron1m) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:9]: Even the point that we have some disagreements here point to the face this needs to be in a different PR.

Let me ask you this way - if you remove this code and throw here, would this PR be useless or have some meaning?

if it has meaning - lets merge it without this and work on this logic in a different PR.

File: packages/functions/src/common/BookAppointmentHelper.ts [hunk 15]
```diff
@@ -47,7 +47,7 @@ export async function bookAppointment(
   appointmentToBook.lastChangeType = BookingChange.BOOKED;
   appointmentToBook.status = AppointmentStatus.BOOKED;
 
-  if ( coordinatorId && donorId === MANUAL_DONOR_ID) {
+  if (donorId === MANUAL_DONOR_ID) {
```
Comment (yaron1m) [inline packages/functions/src/common/BookAppointmentHelper.ts:50]: I suggest you still check here that coordinatorId is not undefined

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts [hunk 16]
```diff
@@ -1,10 +1,19 @@
...
+    const coordinator = await CoordinatorDAL.getCoordinator(callerId);
+
+    if (!coordinator) {
+      throw Error(`User ${callerId} is not an admin`);
+    }
+  }
```
Comment (yaron1m) [inline packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:15]:
```
  if (request.donorId !== MANUAL_DONOR_ID) {
      throw Error(`Coordinator booking for real donots is not supported yet`);
  }
  
  const coordinator = await CoordinatorDAL.getCoordinator(callerId);

  if (!coordinator) {
    throw Error(`User ${callerId} is not an admin`);
  }
```

File: packages/functions/src/common/BookAppointmentHelper.ts (hunk 15 above)
Comment (yedidya03) [reply packages/functions/src/common/BookAppointmentHelper.ts]: done

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts (hunk 16 above)
Comment (yedidya03) [reply packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts]: done

**Comment by yaron1m** [comment]
Looks good to me!

Great job!

File: packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  request: FunctionsApi.CoordinatorBookAppointmentRequest,
+  callerId: string
+): Promise<FunctionsApi.BookAppointmentResponse> {
+  if (request.donorId !== MANUAL_DONOR_ID) {
+    throw Error(`Coordinator booking for real donots is not supported yet`);
+  }
+
+  const coordinator = await CoordinatorDAL.getCoordinator(callerId);
+
+  if (!coordinator) {
```
Comment (yaron1m) [inline packages/functions/src/coordinator/CoordinatorBookAppointmentHandler.ts:15]: What about checking the coordinator has the right permissions for the appointment (i.e. belongs to the right hospital)?

File: packages/functions/src/dal/AdminDataAccessLayer.ts [hunk 18]
```diff
@@ -22,3 +24,37 @@ export async function deleteAdmin(adminId: string) {
   const collection = admin.firestore().collection(Collections.COORDINATORS);
   await collection.doc(adminId).delete();
 }
+
+export async function getAdminDonors(adminId: string): Promise<DbDonor[]> {
```
Comment (yaron1m) [inline packages/functions/src/dal/AdminDataAccessLayer.ts:28]: please rename to getCoordinatorDonors

**Review by yaron1m — COMMENTED** [review]
Please see my follow up comments
